rfcs/SubschemaComposability.md

```diff
@@ -0,0 +1,74 @@
+2. "Run-time" composition in which the composite schema is created dynamically at system start-up.
+3. "Dynamic" composition , in that subschemas are polled by a managing service (or report changes to a managing service), such that the composition process can be triggered at or after system start-up and that the composite schema is dynamically refreshed as the subschemas change (or go offline!).
+
+### Method of Composite Schema Execution
+
+1. `Subschemas as remote GraphQL services`: in certain forms of composition (gateways, gateway schemas, etc.), the execution of a request against a composite schema will yields request(s) against spec-compliant (or even non-spec compliant) GraphQL subservices that are then merged into a single request.
+2. `Single-service composite schema execution`: Often, especially when utilizing "build-time" composition, a single service executes the composite schema without delegation to remote services tied to each subschema. 
+
+The below RFC is meant to be useful across all varieties of schema composition. Because all forms of schema composition require a composite schema, we can speak of "composability" without referring to the timing of composition or the type of execution. However, when relevant, reference will be made to underlying technical factors which affect "composability." Practically speaking, the method of composite schema execution matters more than the timing of composition. If a request against the composite schema must ultimately be translated to request(s) against the still extant source subschemas, then the composition must be -- in some sense -- reversed during execution, an additional strenuous requirement.
+
+# Scalars
+
+1. The specified scalars (Int, Float, String, Boolean, ID) can each be merged across subschemas, because they are identical in all subschemas. 
+2. Custom scalars can be merged across subschemas, _as long as the parsing/serialization of values is performed in the identical manner across all subschemas._
+   - This important condition/qualification may be difficult to verify!
+   - The `specifiedByURL` field may be helpful in this regard.
+
+# Enums
+
+1. Overlapping enum types where the types define identical sets of values can always be merged, as they are identical in all subschemas.
+2. Subschemas with overlapping enum types where the disparate types define different value sets are only sometimes composable.
+   - When using `Single-service composite schema execution`, the enum types can be merged as a union of values from all subschemas.
+   - When using `Subschemas as remote GraphQL services`
+     - If the enum types are used only in output types, the enum types can be merged as a union of values from all subschemas.
+     - If the enum types are used in input types, the enum types can be merged if and only if there is a way to translate, for any given subschema, every enum value that is defined only in the composite schema to an enum value defined for that subschema.
```

Comment:
I briefly posted a version that allowed intersections of enums, but I think that unfortunately we have to fail the composition altogether.

For example, consider two scenarios:

Perhaps a field within a subschema defines an argument called order of enum type OrderEnum with values ASCENDING and DESCENDING. If DESCENDING is filtered out from the composed schema, the composed schema has fewer capabilities than at least one of the subschemas, but all data represented by the subschema is accessible.

Consider another scenario in which a field has an argument called filter of enum type FilterEnum with values ItemTypeOne, ItemTypeTwo etc that are meant to be an exhaustive list of the possible item types returned by this field. If values for this enum are filtered from the enum in the composite schema, then the functionality of the filter argument no longer makes sense, and the whole argument should probably not be included in the composite schema.

Seeing a subset would be surprising in both cases, but I think it would be preferable in the latter case to fail the merge. Unfortunately, I don't see a straightforward way to delineate between the two cases without human knowledge of the behavior of the argument.

Reply:
I didn't quite follow the first example, but your second example makes perfect sense. 👍 
We simply can't have an output enum value which isn't already described by the full schema. Bummer!

Reply:
The problem of remote subschema breaking changes bubbling up to the composed schema is pretty serious — but is it more serious than any remote service breaking change would be for a non-composed GraphQL service?

In any event, I am not clear how subschema metadata would help us navigate the challenge. If a subschema removes a field, for example, what kind of metadata could help?

If a subschema adds a field that clashes with another subschema, presumably the composed schema has no knowledge of the change, so the clash will not actually cause a runtime error.

Standardizing subschema metadata seems useful to me, although I was mostly thinking about in terms of nomenclature for how to perform cross-service resolution at runtime, I.e. key directives…

Reply:
Quick sanity check: did you mean input enum value?